presto-prometheus/src/main/java/io/prestosql/plugin/prometheus/PrometheusClient.java

```diff
@@ -0,0 +1,192 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.prestosql.plugin.prometheus;
+
+import com.google.common.base.Function;
+import com.google.common.base.Supplier;
+import com.google.common.base.Suppliers;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.io.Resources;
+import io.airlift.json.JsonCodec;
+import io.prestosql.spi.type.DoubleType;
+import io.prestosql.spi.type.TimestampType;
+import io.prestosql.spi.type.TypeManager;
+import io.prestosql.spi.type.TypeSignature;
+
+import javax.inject.Inject;
+
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.net.URL;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import static com.google.common.collect.Iterables.transform;
+import static com.google.common.collect.Maps.transformValues;
+import static com.google.common.collect.Maps.uniqueIndex;
+import static java.nio.charset.StandardCharsets.UTF_8;
+import static java.util.Objects.requireNonNull;
+
+public class PrometheusClient
+{
+    // schema name is fixed to "prometheus"
+    private final Supplier<Map<String, Object>> tables;
+    private final URI prometheusURI;
+    protected final String queryChunkSizeDuration;
+    private static TypeManager typeManager;
+    protected static final String METRICS_ENDPOINT = "/api/v1/label/__name__/values";
+
+    @Inject
+    public PrometheusClient(PrometheusConfig config, JsonCodec<Map<String, Object>> metricCodec, TypeManager typeManager)
+            throws URISyntaxException
+    {
+        requireNonNull(config, "config is null");
+        requireNonNull(metricCodec, "metricCodec is null");
+        requireNonNull(typeManager, "typeManager is null");
+
+        tables = Suppliers.memoize(metricsSupplier(metricCodec, getPrometheusMetricsURI(config)));
```

Comment:
Can we rename `tables` to something more clear that it's not actually the list of tables, but rather a `tableSupplier` or something? It took me a bit to realize you weren't listing tables during class initialization and caching them forever.

Reply:
Actually, reading about this more, the docs indicate this will be cached after the first call forever, which we don't want. We'll probably want to not cache by default, and potentially we could have an option to cache, or if the strain is too big, we can have a cache and set it to expire regularly. I'd like to also investigate if we can check if a metric exists some other way besides listing all the metric names.

Reply:
Refactored name. Addressed the caching by setting a short cache, 30 seconds: 13ff799